doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,669 @@
+
+#### Pros
+
+- Conceptually simple.
+- Matches the most common pattern of pageable HTTP APIs allowing e.g.
+  the page token to be a thin wrapper around a "real" page token(s).
+- Manager plugin can remain conceptually stateless, since required state
+  is bundled up in the page token.
+- Page token is (or can be) immutable, aiding thread-safety.
+- Using the page token to determine if more pages are available (e.g.
+  a `None` token means exhausted) allows the server to override the
+  page size, opening up an avenue for optimisation.
+- Destruction hooks on a page token (e.g. destructor of a class-based
+  token) could signal to the manager that the page is no longer
+  required, allowing e.g. cache cleanup.
+
+#### Cons
+
+- Adding more functionality (e.g. previous page) requires additional
+  sibling functions or arguments.
+- Adding metadata (e.g. total results) requires additional sibling
+  functions or arguments.
+- Function invocations to fetch subsequent pages must pass all the same
+  arguments, such that they are valid for the page token. This is
+  error-prone and/or redundant.
+
+### Paging option 3 - Cursor objects / iterators
+
+Cursor objects are an abstraction over some unknown internal paging
+mechanism, which may be a page-token, or some other mechanism.
+This cursor could be an object-oriented wrapper with methods, e.g.
+`.next()`, or a data structure with opaque links, e.g. containing a URL
+to the "next" page.
+
+For example, a simple paging cursor object may look like:
+```python
+class AbstractCursor:
+
+    def getPage(self, pageSize: int) -> List[EntityReference]:
+        # Get next page of refs.
+        ...
+
+    def skip(self, pageSize: int):
+        # Skip over records without fetching.
+        ...
+
+    def remaining(self) -> Union[int, None]:
+        # None means "unsupported".
+        ...
+```
+
+#### Pros
+
+- Retrieving the data can be decoupled from traversal of pages, allowing
+  for lazy network retrieval of desired pages.
+- Additional metadata can be included in the cursor (e.g. remaining
+  pages) in an extensible way.
+- A mutable, stateful cursor object maximizes opportunities for the
+  manager plugin to optimize query caches.
+- Destruction of a cursor object could be used as a signal to the
+  manager that no further pages are required, allowing e.g. cache
+  cleanup.
+
+#### Cons
+
+- Manager plugin must implement yet another interface.
+- Complexity of binding to multiple target languages.
+- A mutable, stateful cursor violates the stateless design principle of
+  interactions with manager plugins. This causes complications, e.g.
+  thread-safety.
+- An immutable cursor object (such that `.next()` returns a new cursor
+  object), could increase optimisation complexity for the manager
+  implementation, e.g. managing query cache sharing, lifetime and
+  updates across multiple "live" cursors.
+- The temptation to directly wrap a "real" DB cursor object in our
```

Comment:
After offline conversation/feedback, added this Con and updated the Outcome slightly in 5d85986d